### DISCUSSION TITLE
Why is `absl::base_internal::invoke` in the internal namespace?

### DISCUSSION THREAD

**Discussion by pfaion** [abseil/abseil-cpp#1477] [General]
Hi everyone, thanks for making and maintaining this great library!

I am using abseil in a C++14 codebase to get support for newer C++ features. I find myself frequently wanting to use `std::invoke`, which is a C++17 feature. Looking into abseil, I found `absl/base/internal/invoke.h`, which implements a version of `std::invoke`. However, this implementation is placed into the internal namespace as `absl::base_internal::invoke`, which AFAIK should not be used from the outside.

1. Is there a specific reason for that?
2. Is there a plan to implement something like `absl::invoke` outside of the internal namespace?

**Comment by derekmauro** [comment]
> Hi everyone, thanks for making and maintaining this great library!
> 
> I am using abseil in a C++14 codebase to get support for newer C++ features. I find myself frequently wanting to use `std::invoke`, which is a C++17 feature. Looking into abseil, I found `absl/base/internal/invoke.h`, which implements a version of `std::invoke`. However, this implementation is placed into the internal namespace as `absl::base_internal::invoke`, which AFAIK should not be used from the outside.
> 
> 1. Is there a specific reason for that?

Abseil's primary purpose is to be used as a dependency by Google's open source C++ projects. This means that Abseil contains the most used pieces of what was once our internal-only codebase. `absl::base_internal::invoke` actually wasn't used at all, so it was not battle-tested by thousands of callers. It was only needed as an implementation detail for a handful of other types. Once you add something to your public API surface it becomes a maintenance cost, so we decided not to do that.

> 2. Is there a plan to implement something like `absl::invoke` outside of the internal namespace?

No. Google-internal is on C++17 now, so we have `std::invoke`. We only add polyfills that we need ourselves.

**Reply by pfaion** [reply]
Thank you for the detailed response!
